Firestore/core/src/local/leveldb_migrations.cc

```diff
@@ -303,6 +307,76 @@ void EnsureCollectionParentsIndex(leveldb::DB* db) {
   transaction.Commit();
 }
 
+/**
+ * Returns a `TargetData` by reading the `targets` table, using the given key
+ * for `query_targets` as a foreign key.
+ */
+util::StatusOr<TargetData> ReadTargetData(
+    const LevelDbQueryTargetKey& query_target_key,
+    const LocalSerializer& serializer,
+    LevelDbTransaction& transaction) {
+  auto target_it = transaction.NewIterator();
+  const auto& target_key = LevelDbTargetKey::Key(query_target_key.target_id());
+  target_it->Seek(target_key);
+  if (!target_it->Valid()) {
+    LOG_WARN("Dangling query-target reference found: seeking %s found %s",
+             DescribeKey(target_key), DescribeKey(target_it));
+    return util::Status(kErrorNotFound,
+                        "Dangling query-target reference found.");
+  }
+
+  StringReader reader{target_it->value()};
+  auto message = Message<firestore_client_Target>::TryParse(&reader);
+  if (!reader.ok()) {
+    HARD_FAIL("Target proto failed to parse: %s", reader.status().ToString());
+  }
+  auto target_data = serializer.DecodeTargetData(&reader, *message);
+  if (!reader.ok()) {
+    HARD_FAIL("Target failed to parse: %s, message: %s",
+              reader.status().ToString(), message.ToString());
+  }
+
+  return target_data;
+}
+
+/**
+ * Migration 7.
+ *
+ * Rewrites targets canonical IDs with new format.
+ */
+void RewriteTargetsCanonicalIds(leveldb::DB* db,
+                                const LocalSerializer& serializer) {
+  LevelDbTransaction transaction(db, "Rewrite Targets Canonical Ids");
+
+  std::string query_targets_prefix = LevelDbQueryTargetKey::KeyPrefix();
+  auto it = transaction.NewIterator();
+  it->Seek(query_targets_prefix);
+  LevelDbQueryTargetKey query_target_key;
+  for (; it->Valid() && absl::StartsWith(it->key(), query_targets_prefix);
+       it->Next()) {
+    HARD_ASSERT(query_target_key.Decode(it->key()),
+                "Failed to decode query_targets key");
+
+    util::StatusOr<TargetData> target_data =
+        ReadTargetData(query_target_key, serializer, transaction);
+    if (!target_data.ok()) {
+      LOG_WARN("Reading target data failed: %s",
+               target_data.status().error_message());
+      continue;
+    }
+
+    auto new_key = LevelDbQueryTargetKey::Key(
+        target_data.ValueOrDie().target().CanonicalId(),
+        target_data.ValueOrDie().target_id());
+
+    transaction.Delete(it->key());
+    std::string empty_buffer;
+    transaction.Put(new_key, empty_buffer);
```

Comment:
We need to write the old data back (`empty_buffer` sounds suspicious).

Reply:
This table always has empty value: https://osscs.corp.google.com/firebase-sdk/firebase-ios-sdk/+/master:Firestore/core/src/local/leveldb_target_cache.cc;l=110

It serves as an association between canonical id and target id.

Reply:
Got it.